File: Python/Problem_007.py
# Author: Christian Seely

import math

# Method to determine if a number n is prime.
def isPrime(n):
    div = 2
    while div <= math.sqrt(n):
        if (n % div) == 0:
            return False
        div += 1
    return True

def problemSeven():
    numPrime, i = 0, 2
    while numPrime < 10001:
        if isPrime(i):
            numPrime += 1
        i += 1
    return i-1

if __name__ == '__main__':
    print(problemSeven())

File: Haskell/Problem_007.hs
-- Author: Christian Seely

-- Note, the prime generator was derived from the paper: https://www.cs.hmc.edu/~oneill/papers/Sieve-JFP.pdf

primeGen :: [Integer]
primeGen = 2 : rest
  where rest = [c | c <- [3..], all ((/=) 0 . mod c) $ takeWhile (\p -> p*p <= c) primeGen]

nthPrime :: Int -> Integer
nthPrime n = primeGen !! n

problemSeven :: Integer
problemSeven = nthPrime 10000

main = print $ problemSeven